uPy/html.py

```diff
@@ -0,0 +1,191 @@
+# Simple HTML parser for PMT to use
+# Curran Hyde
+#
+#   Returns a list of dictionaries. Each dictionary is a form:
+#       
+#       Pick a form by indexing into the list. Each form is a dictionary.
+#      
+#       Dictionary key value pairs are keys and values within that tag line:
+#            ex from html: name="formname" method="POST"
+#            becomes:   dict["name"] -> "formname"
+#                       dict["method"] -> "POST"
+#   Structure of return value:
+#       return value will be a list
+
+def parse_forms(html):
+    currentForm = -1
+    forms = []
+    formOpen = 0 #indicates if <form ...> seen but </form> not seen yet
+
+    #int to str
+    html = html.decode('utf-8')
+
+    # empty condition: give back empty dict
+    if(html == ''):
+        return []
+    
+    # Note: Tags can end on same line they were created on
+
+    # State List
+    # NOT_IN_TAG = 0  # 0 = not in a tag. b'<' not seen yet
+    # IN_TAG = 1      # 1 = inside a tag. b'<' seen but b'>' not seen
+
+    state = 0
+    i = 0
+    print("Entering Form parse loop")
+    while i < len(html):
+        byte = html[i]
+
+        # STATE 0 searching ..
+        if(state==0):
+            if(byte != '<' ):
+                i=i+1
+                continue
+            else:
+                state=1
+                #print("Tag Detected: <")
+                
+        # STATE 1 in a tag
+        if(state==1):
+            # find postion of b'>' starting after b'<' 
+            endpos=html.find('>', i, len(html)-1 )
+            #print("endpos={0}".format(endpos))
+
+           # '>' not found in html. tag starts but does not terminate
+            if(endpos==-1):
+                print("\tError: Tag does not terminate. No >")
+                #print(html[i:len(html)]) #print out the rest of the html for debugging
+                return []
+            
+            # complete tag found '<' and '>'
+            else:
+                # pull out '<' through '>'
+                tag=html[ i:endpos+1 ]
```

Comment:
this is nitpicky but remove the spaces before and after the [ ] and choose whether you want to have spaces before and after every '=' or not spaces, you change it up all over and sometimes mix and match